core/src/main/java/io/questdb/griffin/engine/table/AbstractDeferredTreeSetRecordCursorFactory.java

```diff
@@ -59,44 +59,47 @@ public AbstractDeferredTreeSetRecordCursorFactory(
         // we could pass all symbols to factory to resolve, but this would lead to
         // creating Strings for symbols that we may be able to avoid doing so
 
-        final int nKeyValues = keyValues.size();
+        final int nKeyValues = keyValueFuncs.size();
         final IntHashSet symbolKeys = new IntHashSet(nKeyValues);
-        CharSequenceHashSet deferredSymbols = null;
+        IntHashSet deferredSymbolKeys = null;
+        ObjList<Function> deferredFuncs = null;
 
         for (int i = 0; i < nKeyValues; i++) {
-            CharSequence symbol = keyValues.get(i);
-            int symbolKey = symbolMapReader.keyOf(symbol);
+            Function symbolFunc = keyValueFuncs.get(i);
+            int symbolKey = symbolFunc.isRuntimeConstant()
+                    ? SymbolTable.VALUE_NOT_FOUND
+                    : symbolMapReader.keyOf(symbolFunc.getStr(null));
             if (symbolKey == SymbolTable.VALUE_NOT_FOUND) {
-                if (deferredSymbols == null) {
-                    deferredSymbols = new CharSequenceHashSet();
+                if (deferredFuncs == null) {
+                    deferredFuncs = new ObjList<>();
+                    deferredSymbolKeys = new IntHashSet();
                 }
-                deferredSymbols.add(Chars.toString(symbol));
+                deferredFuncs.add(symbolFunc);
             } else {
                 symbolKeys.add(TableUtils.toIndexKey(symbolKey));
             }
         }
 
         this.columnIndex = columnIndex;
         this.symbolKeys = symbolKeys;
-        this.deferredSymbols = deferredSymbols;
+        this.deferredSymbolKeys = deferredSymbolKeys;
+        this.deferredSymbolFuncs = deferredFuncs;
     }
 
     @Override
     protected AbstractDataFrameRecordCursor getCursorInstance(
             DataFrameCursor dataFrameCursor,
             SqlExecutionContext executionContext
     ) throws SqlException {
-        if (deferredSymbols != null && deferredSymbols.size() > 0) {
+        if (deferredSymbolFuncs != null) {
```

Comment:
I'd rather have 1 map to lookup in row by row evaluation instead of 2.
I understand we need to keep `symbolKeys` and top them up with `deferredSymbolKeys` on every `getCursorInstance` but I think it's better to store the prepared results in 1 map instead of 2.

Reply:
There should be only one of the maps in use effectively. The only case when both of them would contain values is a mix of constant literals (`in ('known_symbol', 'unknown_symbol')`) used for the key column filter. I didn't like the idea of having to rebuild the map on each query execution in the situation when the query doesn't use bind variables, so decided to keep two separate maps: "static" and "dynamic" one. WDTY?

Reply:
I only now understood how these interrelated lists are used around, so it' fine to have 2 list of keys here.

Reply:
Let me add some comments to improve the code clarity then.